Remove mock from test_cli

exarkun opened the pull request:
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3528

sajith commented on src/allmydata/scripts/runner.py:
```diff
@@ -223,21 +262,43 @@ def write_coverage_data():
+
+    :return: A ``Deferred`` that fires when the run is complete.
+    """
+    _setup_coverage(reactor, argv)
```
Why is additional setup for coverage needed here?  Could we document the reasons?

exarkun replied:
added some more words in 9958236c3

sajith reviewed (approved):
Looks good to me, except for a couple of comments about possibly minor things I have not understood.

sajith commented on src/allmydata/scripts/runner.py:
```diff
@@ -119,7 +137,7 @@ def parse_or_exit_with_explanation(argv, stdout=sys.stdout):
         except Exception:
             msg = repr(e)
-        print("%s:  %s\n" % (sys.argv[0], quote_output(msg, quotemarks=False)), file=stdout)
+        print("%s:  %s\n" % (argv[0], quote_output(msg, quotemarks=False)), file=stdout)
```
This is what has confused me:

https://github.com/tahoe-lafs/tahoe-lafs/blob/6e152daf05f01191c88a1515f313c060baacdee8/src/allmydata/scripts/runner.py#L132-L135

What does it do?

exarkun replied:
Hopefully explained in c9b3ccedb

sajith reviewed (dismissed):
Seems that tests fail on Windows.  Can we address that?

meejah commented on src/allmydata/scripts/runner.py:
```diff
@@ -180,19 +222,44 @@ def run():
+    )
+
+
+def _setup_coverage(reactor, argv):
```
I wonder if "coverage" has a better way yet? (I feel like I've explored "why isn't sub-process coverage working" too many times already though :/ )

exarkun replied:
Last time I looked at it, it didn't, and the ticket and PR I filed against it have seen zero response...

meejah commented on src/allmydata/scripts/runner.py:
```diff
@@ -223,21 +290,43 @@ def write_coverage_data():
+
+    :param reactor: The reactor to use.  This implementation largely ignores
+        this and lets the rest of the implementation pick its own reactor.
+        Oops.
```
So what *does* this function do differently then? Can it just be merged with `run()`?

exarkun replied:
It serves as a starting point to start respecting the reactor parameter, at least.  From this point inward we *could* be using that reactor, if we wanted.  `run` is the thing that picks a value for that reactor.

meejah reviewed (approved):
LGTM, some inline comments

exarkun commented:
more than a few small changes since the last review, so I think it makes sense to have this reviewed again - or at least the changes since the last review.

itamarst commented on src/allmydata/scripts/runner.py:
```diff
@@ -9,6 +9,7 @@
 
 import os, sys
 from six.moves import StringIO
+from past.builtins import unicode
```
Given we're in post-Python-3-port world, should just be using `str()`.

itamarst commented on src/allmydata/scripts/runner.py:
```diff
@@ -115,23 +116,75 @@ def postOptions(self):
-        # may be unicode, in which case it will print funny. Once we're on
-        # Python 3 we can just drop the ensure_str().
-        print(six.ensure_str("%s:  %s\n" % (sys.argv[0], e)), file=stdout)
+        exc_str = str(e)
```
:scream: but I guess I can see why this would happen.

itamarst commented on src/allmydata/scripts/tahoe_run.py:
```diff
@@ -212,10 +212,11 @@ def run(config):
+    twistd_args = ["--nodaemon", "--rundir", basedir]
+    if sys.platform != "win32":
+        pidfile = get_pidfile(basedir)
+        twistd_args.extend(["--pidfile", pidfile])
```
Where did the PID file come from before if this wasn't set?

exarkun replied:
The code used to rely on the default pidfile behavior from twistd.  That implicit behavior depended on interactions with `os.chdir` to get a predictable path which made the tests more complicated (eg they previously monkey-patched `os.chdir` to keep some kind of control over what was happening).
